File: src/main/java/demos/oop/order/refactoring/Order.java
package demos.oop.order.refactoring;

import java.util.HashMap;
import java.util.List;

class Order {
    private HashMap<String, OrderLine> linesById;
    private String orderId;

    public Order(String orderId) {
        this.orderId = orderId;
        this.linesById = new HashMap<String, OrderLine>();
    }

    public void add(String productId, int amount) {
        if (this.linesById.containsKey(productId)) {
            this.linesById.get(productId).add(amount);
        } else {
            this.linesById.put(productId, new OrderLine(productId, amount));
        }
    }

    public List<OrderLine> getLines() {
        return List.copyOf(this.linesById.values());
    }
    
    public String getOrderId() {
        return orderId;
    }
}

File: src/main/scala/demos/fp/order/refactoring/Order.scala
package demos.fp.order.refactoring

import scala.collection.immutable.HashMap

class Order private(private val linesById: HashMap[String, OrderLine] = HashMap(), val orderId: String):
    def add(productId: String, amount: Int): Order =
        val existingOrderLine = this.linesById.get(productId)
        val updatedLines = existingOrderLine match
            case None =>
                val newLine = OrderLine(productId, amount)
                this.linesById + (productId -> newLine)
            case Some(orderLine) =>
                val updatedLine = orderLine.add(amount)
                this.linesById + (productId -> updatedLine)
        new Order(linesById = updatedLines, orderId=this.orderId)

    def lines: List[OrderLine] = List.from(this.linesById.values)

    override def toString: String = s"Order(lines=${this.lines}, orderId=${this.orderId})"


object Order:
    def apply(orderId: String): Order = new Order(orderId = orderId)
